src/asyncresults.jl

```diff
@@ -0,0 +1,242 @@
+[Asynchronous Command Processing](https://www.postgresql.org/docs/10/libpq-async.html).
+
+The `throw_error` option only determines whether to throw errors when handling the new
+[`Result`](@ref)s; the `Task` may error for other reasons related to processing the
+asynchronous loop.
+
+The result returned from the `Task` will be the [`Result`](@ref) of the last query run (the
+only query if using parameters).
+Any errors produced by the queries will be thrown together in a `CompositeException` by
+`@sync`.
+"""
+function consume(async_result::AsyncResult; throw_error=true)
+    @async begin
+        try
+            debug(LOGGER, "getting the socket fd")
+            pqfd = socket(async_result.jl_conn)
+            try
+                debug(LOGGER, "making an FDWatcher")
+                watcher = FDWatcher(pqfd, true, false)  # readable, not writeable
+                try
+                    last_result = @sync begin
+                        result = nothing
+                        debug(LOGGER, "beginning the waiting loop")
+                        while true
+                            debug(LOGGER, "waiting on the fd")
+                            wait(watcher)
+                            debug(LOGGER, "consuming input")
+                            success = libpq_c.PQconsumeInput(async_result.jl_conn.conn) == 1
+                            !success && error(error_message(async_result.jl_conn))
+                            debug(LOGGER, "checking isbusy")
+                            while libpq_c.PQisBusy(async_result.jl_conn.conn) == 0
+                                debug(LOGGER, "getting a result")
+                                result_ptr = libpq_c.PQgetResult(async_result.jl_conn.conn)
+                                result_ptr == C_NULL && @goto finished
+                                debug(LOGGER, "handling the non-null result")
+                                result = @async handle_result(Result(
+                                    result_ptr, async_result.jl_conn;
+                                    async_result.result_kwargs[]...
+                                ); throw_error=throw_error)
+                            end
+                        end
+                        @label finished
+                        debug(LOGGER, "finished the sync block")
+                        result
+                    end
+
+                    return fetch(last_result)
+                finally
+                    close(watcher)
+                end
+            finally
+                # I should maybe close pqfd here but I don't think that's possible since
+                # it's a file descriptor not a handle and it's not technically open
+            end
+        finally
+            # ensure the connection is unlocked
+            _ensure_unlocked(async_result)
+            debug(LOGGER, "finished the async block")
+        end
+    end
+end
+
+function cancel(async_result::AsyncResult)
```

Comment:
I feel like maybe we should be having a `async_result.cancelled` atomic?
And that we can do the `if !atomic_cas!(async_result.cancelled, false, true)`
to guard the body of `cancel` from being called twice at the same time.

Reply:
Cancelling is a thing you do to a connection, not a result though. I could add a check of `conn_locked` I guess though.

Reply:
Right now I get it.
Shall we include a comment here saying: 
```suggestion
function cancel(async_result::AsyncResult)
    # Just set the `should_cancel` flag, the actually cancellation will be triggered in the main loop of _consume
    # which will call `_cancel` on the conection  
```